### PR TITLE
Add a VC707 chiplink slave target

### PR DESCRIPTION
**Pull request by terpstra**
This design can be attached to a HiFive Unleashed board to add PCIe functionality.

### CODE REVIEW

File: src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala [hunk 1]
```diff
@@ -0,0 +1,233 @@
+// See LICENSE for license details.
+
+// ChipLink to GPIO
```
Comment (yunsup) [inline src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala]: delete comment

File: src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+import sifive.fpgashells.shell.xilinx.vc707shell.{VC707Shell,HasPCIe,HasDDR3,HasVC707ChipLink}
+import sifive.fpgashells.ip.xilinx.{IOBUF,vc707_sys_clock_mmcm3}
+import sifive.fpgashells.devices.xilinx.xilinxvc707mig._
+import sifive.fpgashells.devices.xilinx.xilinxvc707pciex1._
+
+import sifive.freedom.unleashed.u500vc707devkit.FreedomUVC707Config
+
+import freechips.rocketchip.util.{ElaborationArtefacts,ResetCatchAndSync}
+
+// ChipLink
```
Comment (yunsup) [inline src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala]: this is obvious, I'd delete the comment

File: src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  def apply(): BasePin = {
+    new BasePin()
+  }
+}
+
+//-------------------------------------------------------------------------
+// IOFPGAChip
+//-------------------------------------------------------------------------
+
+class ShadowRAMHack(implicit p: Parameters) extends LazyModule
```
Comment (yunsup) [inline src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala]: What is this used for?  Maybe leave comment?

File: src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    io.chiplink <> link.module.io.port
+    link.module.io.c2b_clk := clock
+    link.module.io.c2b_rst := ResetCatchAndSync(clock, reset || timer.orR || !io.rxlocked)
+
+    io.gpio <> gpio.module.io.port
+  }
+}
+
+class IOFPGAChip(implicit override val p: Parameters) extends VC707Shell
+//  with HasPCIe
```
Comment (yunsup) [inline src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala]: delete comment

File: src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        TLC =  AddressSet.misaligned(0x60000000L,   0x20000000L) ++                 // local memory behind L2 [1.5GB, 2GB)
+               AddressSet.misaligned(0x80000000L,   0x2000000000L - 0x80000000L) ++ // Aloe DDR               [  2GB, 128GB)
+               AddressSet.misaligned(0x3000000000L, 0x1000000000L),                 // local memory behind L2 [192GB, 256GB)
+        syncTX = true
+  )
+  val localRoute = AddressSet.misaligned(0x40000000L, 0x20000000L) ++               // local MMIO             [  1GB, 1.5GB)
+                   AddressSet.misaligned(0x2000000000L, 0x1000000000L)              // local MMIO             [128GB, 192GB)
+  val gpioParams = GPIOParams(address = BigInt(0x2400000000L), width = 4)
+
+  //ChipLink skew RX clock
```
Comment (yunsup) [inline src/main/scala/unleashed/vc707/iofpga/FPGAChip.scala]: space after //

**Comment by terpstra** [comment]
I think:
unleashed/vc707/iofgpa
unleashed/vc707/devkit
unleashed/xxx/iofpga
... works just as well.

**Comment by terpstra** [comment]
The produced FPGA image was confirmed able to successfully host quake2.

**Pull request merged**